File: plugins/togetherai-skills/skills/together-audio/scripts/tts_generate.ts
#!/usr/bin/env -S npx tsx
/**
 * Together AI text-to-speech examples with the TypeScript SDK.
 *
 * Demonstrates:
 * - REST file generation
 * - Streaming HTTP generation
 * - Voice discovery
 *
 * Usage:
 *   npx tsx tts_generate.ts --mode rest --text "Hello world" --output speech.mp3
 *   npx tsx tts_generate.ts --mode stream --text "Hello world" --output speech_stream.pcm
 *   npx tsx tts_generate.ts --mode voices
 *
 * Requirements:
 *   npm install together-ai
 *   export TOGETHER_API_KEY=your_key
 */

import Together from "together-ai";
import { createWriteStream, promises as fs } from "fs";
import { Readable } from "stream";
import { pipeline } from "stream/promises";

const client = new Together();

type Mode = "rest" | "stream" | "voices";

interface ParsedArgs {
  mode: Mode;
  text: string;
  output: string;
  model: string;
  voice: string;
  responseFormat: "mp3" | "wav";
  responseEncoding: "pcm_f32le" | "pcm_s16le" | "pcm_mulaw" | "pcm_alaw";
  language?: string;
  sampleRate?: number;
  alignment: "none" | "word";
  segment: "sentence" | "immediate" | "never";
}

function parseArgs(): ParsedArgs {
  const raw = process.argv.slice(2);
  const get = (flag: string): string | undefined => {
    const index = raw.indexOf(flag);
    if (index === -1) return undefined;
    return raw[index + 1];
  };

  const mode = (get("--mode") ?? "rest") as Mode;
  const responseFormat = (get("--response-format") ?? "mp3") as ParsedArgs["responseFormat"];
  const responseEncoding = (get("--response-encoding") ?? "pcm_s16le") as ParsedArgs["responseEncoding"];
  const alignment = (get("--alignment") ?? "none") as ParsedArgs["alignment"];
  const segment = (get("--segment") ?? "sentence") as ParsedArgs["segment"];
  const sampleRate = get("--sample-rate");

  if (!["rest", "stream", "voices"].includes(mode)) {
    console.error("Expected --mode to be one of: rest, stream, voices");
    process.exit(1);
  }

  return {
    mode,
    text: get("--text") ?? "Today is a wonderful day to build something people love!",
    output: get("--output") ?? "speech.mp3",
    model: get("--model") ?? "canopylabs/orpheus-3b-0.1-ft",
    voice: get("--voice") ?? "tara",
    responseFormat,
    responseEncoding,
    language: get("--language"),
    sampleRate: sampleRate ? Number(sampleRate) : undefined,
    alignment,
    segment,
  };
}

async function writeBodyToFile(body: unknown, outputFile: string): Promise<void> {
  if (!body) {
    throw new Error("Expected a response body");
  }

  if (typeof (body as NodeJS.ReadableStream).pipe === "function") {
    await pipeline(body as NodeJS.ReadableStream, createWriteStream(outputFile));
    return;
  }

  if (typeof (body as ReadableStream<Uint8Array>).getReader === "function") {
    const nodeStream = Readable.fromWeb(body as ReadableStream<Uint8Array>);
    await pipeline(nodeStream, createWriteStream(outputFile));
    return;
  }

  throw new Error("Unsupported response body type");
}

function parseStreamingPayloads(chunk: unknown): Array<Record<string, unknown>> {
  if (chunk && typeof chunk === "object" && "type" in (chunk as Record<string, unknown>)) {
    return [chunk as Record<string, unknown>];
  }

  const text =
    typeof chunk === "string"
      ? chunk
      : chunk instanceof Uint8Array
        ? Buffer.from(chunk).toString("utf8")
        : Buffer.isBuffer(chunk)
          ? chunk.toString("utf8")
          : "";

  if (!text) return [];

  const payloads: Array<Record<string, unknown>> = [];
  for (const line of text.split("\n")) {
    if (!line.startsWith("data:")) continue;
    const data = line.slice(5).trim();
    if (!data || data === "[DONE]") continue;
    try {
      payloads.push(JSON.parse(data) as Record<string, unknown>);
    } catch {
      // Ignore malformed SSE lines.
    }
  }
  return payloads;
}

async function generateRest(args: ParsedArgs): Promise<void> {
  const payload: Record<string, unknown> = {
    model: args.model,
    input: args.text,
    voice: args.voice,
    response_format: args.responseFormat,
  };
  if (args.language) payload.language = args.language;
  if (typeof args.sampleRate === "number") payload.sample_rate = args.sampleRate;

  const response = await client.audio.speech.create(payload);
  await writeBodyToFile(response.body, args.output);
  console.log(`Saved ${args.responseFormat} audio to ${args.output}`);
}

async function generateStream(args: ParsedArgs): Promise<void> {
  const payload: Record<string, unknown> = {
    model: args.model,
    input: args.text,
    voice: args.voice,
    stream: true,
    response_format: "raw",
    response_encoding: args.responseEncoding,
    alignment: args.alignment,
    segment: args.segment,
  };
  if (args.language) payload.language = args.language;
  if (typeof args.sampleRate === "number") payload.sample_rate = args.sampleRate;

  const response = await client.audio.speech.create(payload);
  const rawChunks: Buffer[] = [];

  for await (const chunk of response as AsyncIterable<unknown>) {
    for (const event of parseStreamingPayloads(chunk)) {
      if (event.type === "conversation.item.audio_output.delta" && typeof event.delta === "string") {
        rawChunks.push(Buffer.from(event.delta, "base64"));
      }
      if (event.type === "conversation.item.word_timestamps") {
        console.log(JSON.stringify(event));
      }
    }
  }

  await fs.writeFile(args.output, Buffer.concat(rawChunks));
  console.log(`Saved raw streaming audio to ${args.output}`);
}

async function listVoices(): Promise<void> {
  const response = await client.audio.voices.list();
  for (const modelVoices of response.data ?? []) {
    console.log(`Model: ${modelVoices.model}`);
    for (const voice of modelVoices.voices ?? []) {
      console.log(`  - ${voice.name}`);
    }
  }
}

async function main(): Promise<void> {
  const args = parseArgs();

  if (args.mode === "voices") {
    await listVoices();
    return;
  }

  if (args.mode === "rest") {
    await generateRest(args);
    return;
  }

  await generateStream(args);
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});

File: plugins/togetherai-skills/skills/together-audio/scripts/tts_generate.py
#!/usr/bin/env python3
"""
Together AI text-to-speech examples with the Python v2 SDK.

Demonstrates:
- REST file generation
- Streaming HTTP generation
- Raw PCM byte output
- Voice discovery

Usage:
    python tts_generate.py --mode rest --text "Hello world" --output speech.mp3
    python tts_generate.py --mode stream --text "Hello world" --output speech_stream.wav
    python tts_generate.py --mode raw --text "Hello world" --output speech_raw.pcm
    python tts_generate.py --mode voices

Requirements:
    uv pip install "together>=2.0.0"
    export TOGETHER_API_KEY=your_key
"""

from __future__ import annotations

import argparse
import base64
import struct
from pathlib import Path

from together import Together
from together import Omit

client = Together()


def _write_wav(pcm_data: bytes, output_file: Path, sample_rate: int = 24000) -> None:
    """Wrap raw PCM s16le bytes in a WAV header and write to file."""
    num_channels = 1
    bits_per_sample = 16
    byte_rate = sample_rate * num_channels * bits_per_sample // 8
    block_align = num_channels * bits_per_sample // 8
    data_size = len(pcm_data)

    with open(output_file, "wb") as f:
        f.write(b"RIFF")
        f.write(struct.pack("<I", 36 + data_size))
        f.write(b"WAVE")
        f.write(b"fmt ")
        f.write(struct.pack("<IHHIIHH", 16, 1, num_channels, sample_rate, byte_rate, block_align, bits_per_sample))
        f.write(b"data")
        f.write(struct.pack("<I", data_size))
        f.write(pcm_data)


def generate_rest(
    text: str,
    output_file: Path,
    model: str,
    voice: str,
    response_format: str,
    language: str | None,
    sample_rate: int | None,
) -> None:
    """Generate a complete audio file over HTTP."""
    payload: dict[str, object] = {
        "model": model,
        "input": text,
        "voice": voice,
        "response_format": response_format,
    }
    if language:
        payload["language"] = language
    if sample_rate is not None:
        payload["sample_rate"] = sample_rate

    response = client.audio.speech.create(**payload)
    response.write_to_file(str(output_file))
    print(f"Saved {response_format} audio to {output_file}")


def generate_stream(
    text: str,
    output_file: Path,
    model: str,
    voice: str,
    response_encoding: str,
    language: str | None,
    sample_rate: int | None,
    alignment: str,
    segment: str,
) -> None:
    """Generate streaming audio and save it as a WAV file."""
    extra: dict[str, object] = {
        "alignment": alignment,
        "segment": segment,
    }
    response = client.audio.speech.create(
        model=model,
        input=text,
        voice=voice,
        stream=True,
        response_format="raw",
        response_encoding=response_encoding,
        language=language if language else Omit(),
        sample_rate=sample_rate if sample_rate is not None else Omit(),
        extra_body=extra,
    )
    raw_chunks: list[bytes] = []
    for chunk in response:
        if chunk.type == "conversation.item.audio_output.delta":
            raw_chunks.append(base64.b64decode(chunk.delta))
        elif chunk.type == "conversation.item.word_timestamps":
            print(f"Timestamps: {chunk.model_extra}")

    pcm_data = b"".join(raw_chunks)
    _write_wav(pcm_data, output_file, sample_rate=sample_rate or 24000)
    print(f"Saved streaming audio to {output_file}")


def generate_raw_bytes(
    text: str,
    output_file: Path,
    model: str,
    voice: str,
    response_encoding: str,
    language: str | None,
    sample_rate: int | None,
) -> None:
    """Request raw PCM bytes and save them directly."""
    payload: dict[str, object] = {
        "model": model,
        "input": text,
        "voice": voice,
        "response_format": "raw",
        "response_encoding": response_encoding,
    }
    if language:
        payload["language"] = language
    if sample_rate is not None:
        payload["sample_rate"] = sample_rate

    response = client.audio.speech.create(**payload)
    response.write_to_file(str(output_file))
    print(f"Saved raw audio bytes to {output_file}")


def list_voices() -> None:
    """List every voice returned by the voices API."""
    response = client.audio.voices.list()
    for model_voices in response.data:
        print(f"Model: {model_voices.model}")
        for voice in model_voices.voices:
            print(f"  - {voice.name}")


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Together AI TTS examples")
    parser.add_argument(
        "--mode",
        choices=("rest", "stream", "raw", "voices"),
        default="rest",
        help="Workflow to run",
    )
    parser.add_argument(
        "--text",
        default="Today is a wonderful day to build something people love!",
        help="Input text",
    )
    parser.add_argument(
        "--output",
        default="speech.mp3",
        help="Output file path",
    )
    parser.add_argument(
        "--model",
        default="canopylabs/orpheus-3b-0.1-ft",
        help="TTS model",
    )
    parser.add_argument(
        "--voice",
        default="tara",
        help="Voice identifier",
    )
    parser.add_argument(
        "--response-format",
        choices=("mp3", "wav"),
        default="mp3",
        help="Output format for REST mode",
    )
    parser.add_argument(
        "--response-encoding",
        choices=("pcm_f32le", "pcm_s16le", "pcm_mulaw", "pcm_alaw"),
        default="pcm_s16le",
        help="Raw audio encoding for streaming or raw-byte modes",
    )
    parser.add_argument(
        "--language",
        default=None,
        help="Optional language code such as en, fr, or es",
    )
    parser.add_argument(
        "--sample-rate",
        type=int,
        default=None,
        help="Optional output sample rate in Hz",
    )
    parser.add_argument(
        "--alignment",
        choices=("none", "word"),
        default="none",
        help="Streaming alignment mode",
    )
    parser.add_argument(
        "--segment",
        choices=("sentence", "immediate", "never"),
        default="sentence",
        help="Streaming segmentation mode",
    )
    return parser.parse_args()


def main() -> None:
    args = parse_args()
    output_file = Path(args.output)

    if args.mode == "voices":
        list_voices()
        return

    if args.mode == "rest":
        generate_rest(
            text=args.text,
            output_file=output_file,
            model=args.model,
            voice=args.voice,
            response_format=args.response_format,
            language=args.language,
            sample_rate=args.sample_rate,
        )
        return

    if args.mode == "stream":
        generate_stream(
            text=args.text,
            output_file=output_file,
            model=args.model,
            voice=args.voice,
            response_encoding=args.response_encoding,
            language=args.language,
            sample_rate=args.sample_rate,
            alignment=args.alignment,
            segment=args.segment,
        )
        return

    generate_raw_bytes(
        text=args.text,
        output_file=output_file,
        model=args.model,
        voice=args.voice,
        response_encoding=args.response_encoding,
        language=args.language,
        sample_rate=args.sample_rate,
    )


if __name__ == "__main__":
    main()
